src/mechanics/mesh/MeshGmsh.cpp

```diff
@@ -0,0 +1,375 @@
+#include "MeshGmsh.h"
+
+#include "base/Exception.h"
+#include "mechanics/interpolation/InterpolationTrussLinear.h"
+#include "mechanics/interpolation/InterpolationTriangleLinear.h"
+#include "mechanics/interpolation/InterpolationQuadLinear.h"
+
+#include <array>
+#include <climits>
+#include <algorithm>
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
+
+
+// Helper structs (cpp only)
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+struct GmshHeader
+{
+    double version;
+    bool isBinary;
+    int doubleSize;
+};
+
+struct GmshNode
+{
+    int id = 0;
+    std::array<double, 3> coordinates;
+};
+
+struct GmshElement
+{
+    int id = 0;
+    int type = 0;
+    std::vector<int> tags;
+    std::vector<int> nodes;
+};
+
+struct GmshPhysicalNames
+{
+    int id = 0;
+    int dimension = 0;
+    std::string name;
+};
+
+
+struct GmshFileContent
+{
+    GmshHeader header;
+    int minNodeId = INT_MAX;
+    int maxNodeId = INT_MIN;
```

Comment:
For the mapping during element creation. Gmsh does not guarantee contiguous or ordered IDs. The created `NodeSimple` doesn't know about its gmsh ID. So you can use a map with the ID and a pointer to the created node or a vector with pointers and an easily reproducible mapping.
Let's say you have 3 Nodes with the IDs 1,4,6 and use a densely populated vector with size 3. How do you know that ID 1 is at `[0]`, 4 at `[1]` and 6 at `[2]`. So what we do is creating an oversized vector with the size maxID-minID. In this case you get some vector content like this:

~~~
[0] -> Pointer to node with ID = 1
[1] -> nullptr
[2] -> nullptr
[3] -> Pointer to node with ID = 4
[4] -> nullptr
[5] -> Pointer to node with ID = 6
~~~

Benefit of this approach is that generation and lookup are constant time, while the map gets slower with growing N. Drawback is a potentially higher memory consumption.

Reply:
Looking at the code, you set this variable once (l. 105), but never read it. So it can be removed imho.

Reply:
both IDs are used in the line

~~~cpp
std::vector<NodeSimple*> nodePtrVec(fileContent.maxNodeId - fileContent.minNodeId + 1, nullptr);
~~~

Since the vector isn't filled with anything yet, we can't get the first members Id to get rid of minId :wink: